src/cmd.zig

```diff
@@ -1860,6 +1944,85 @@ fn startGossip(
     return service;
 }
 
+fn replayDependencies(
+    allocator: std.mem.Allocator,
+    epoch: sig.core.Epoch,
+    loaded_snapshot: *sig.accounts_db.snapshot.LoadedSnapshot,
+    bank_fields: *const sig.core.BankFields,
+    app_base: *const AppBase,
+    ledger: UnifiedLedger,
+    epoch_context_manager: *sig.adapter.EpochContextManager,
+) !replay.Dependencies {
+    const epoch_stakes_map = &loaded_snapshot.collapsed_manifest.bank_extra.versioned_epoch_stakes;
+    const epoch_stakes = epoch_stakes_map.get(epoch) orelse
+        return error.EpochStakesMissingFromSnapshot;
+
+    const feature_set = try sig.replay.service.getActiveFeatures(
+        allocator,
+        loaded_snapshot.accounts_db.accountReader().forSlot(&bank_fields.ancestors),
+        bank_fields.slot,
+    );
+
+    const root_slot_constants = try sig.core.SlotConstants.fromBankFields(
+        allocator,
+        bank_fields,
+        feature_set,
+    );
+    errdefer root_slot_constants.deinit(allocator);
+
+    const lt_hash = if (loaded_snapshot.collapsed_manifest.bank_extra.accounts_lt_hash) |lt_hash|
+        sig.core.LtHash{ .data = lt_hash }
+    else
+        null;
+
+    var root_slot_state = try sig.core.SlotState.fromBankFields(allocator, bank_fields, lt_hash);
+    errdefer root_slot_state.deinit(allocator);
+
+    return sig.replay.service.Dependencies{
+        .allocator = allocator,
+        .logger = .from(app_base.logger),
+        .my_identity = .fromPublicKey(&app_base.my_keypair.public_key),
+        .vote_identity = .fromPublicKey(&app_base.my_keypair.public_key),
+        .exit = app_base.exit,
+        .account_store = loaded_snapshot.accounts_db.accountStore(),
+        .ledger = .{
+            .db = ledger.db.*,
+            .reader = ledger.reader,
+            .writer = ledger.result_writer,
+        },
+        .epoch_schedule = bank_fields.epoch_schedule,
+        .slot_leaders = epoch_context_manager.slotLeaders(),
+        .root = .{
+            .slot = bank_fields.slot,
+            .constants = root_slot_constants,
+            .state = root_slot_state,
+        },
+        .current_epoch = epoch,
+        .current_epoch_constants = try .fromBankFields(
+            bank_fields,
+            try epoch_stakes.current.convert(allocator, .delegation),
+        ),
+        .hard_forks = try bank_fields.hard_forks.clone(allocator),
+    };
+}
+
+fn consensusDependencies(
+    allocator: std.mem.Allocator,
+    gossip_table: ?*sig.sync.RwMux(sig.gossip.GossipTable),
+) !replay.service.ConsensusState.Dependencies {
+    const senders: sig.replay.ConsensusState.Senders = try .create(allocator);
+    defer senders.destroy();
+
```

Comment:
I think using `defer` here (and also with receivers) is invalid. You probably want `errdefer`? As this will destroy the channels on a successful return, leaving the returned handles invalid.

Reply:
95776ba99294985e7d4e77f5db7ef8c99cbdb0b9